navigation/differencers.h

```diff
@@ -0,0 +1,110 @@
+/*
+ * This header contains small, functor-type classes for calculating the
+ * differences between images, used by PerfectMemory.
+ */
+
+#pragma once
+
+// Standard C includes
+#include <cmath>
+#include <cstdlib>
+
+// Standard C++ includes
+#include <algorithm>
+#include <vector>
+
+// OpenCV
+#include <opencv2/opencv.hpp>
+
+namespace BoBRobotics {
+namespace Navigation {
+//------------------------------------------------------------------------
+// BoBRobotics::Navigation::AbsDiff
+//------------------------------------------------------------------------
+/*!
+ * \brief For calculating the mean absolute difference between images
+ *
+ * Can be passed to PerfectMemory as a template parameter.
+ */
+class AbsDiff
+{
+public:
+    AbsDiff(const int size)
+      : m_Size(size)
+    {}
+
+    inline void calculateDifference(const cv::Mat &image1, const cv::Mat &image2, cv::Mat &differenceImage)
+    {
+        cv::absdiff(image1, image2, differenceImage);
+        m_Ptr = differenceImage.data;
+    }
+
+    inline uint8_t *begin()
```

Comment:
The function could be made const, but you can't make the pointer const, because ``std::transform`` is using it as an iterator (and thus incrementing it)

Reply:
With const there (rather than ``uint8_t * const begin()``) you can still increment the pointer, you just can't fuck with the thing it's pointing at - which in this case is I think what you want!

Reply:
Ah yeah. I can never remember the rule with adding consts in various places for pointer types...

Reply:
Me neither - there's some spiral rule that doesn't help me at all